### ISSUE TITLE
Could not distinguish foreground and background

### ISSUE THREAD

**Issue by VoyagerXvoyagerx** [yecy749/GSNet#4]
Hi,

I tested a few images on the DIOR and NWPU-VHR-10 datasets using the pretrained model, but I found that the model could not distinguish the foreground and background in any of the images I tested.

Although the training dataset LandDiscover50k includes the DIOR dataset, the performance was still unsatisfactory.

Could you provide some explanations for this issue? Additionally, I have included the code I used for inference and visualization.

Examples are provided below.

Thank you for your assistance.

#### examples in DIOR dataset (using the same class name as LandDiscover)

![Image](https://github.com/user-attachments/assets/34356920-efae-4b1c-8e30-1bbe3374f1a8)
**00252.jpg**

![Image](https://github.com/user-attachments/assets/2dab322f-a664-4268-a572-367386972eb6)
**05300.jpg**

#### examples in NWPU-VHR-10 dataset (use the class names in nwpu dataset with a `background` class)

![Image](https://github.com/user-attachments/assets/8dddd239-5b5f-48fe-a7fd-8426641a1aff)
**019.jpg**

![Image](https://github.com/user-attachments/assets/bfa6bf2c-718b-49da-acd6-33ff60c7a12a)
**132.jpg**

#### Codes inference_gsnet.py
```python
import os
import cv2
import numpy as np
import torch
from detectron2.config import get_cfg
from detectron2.checkpoint import DetectionCheckpointer
from detectron2.engine import default_setup
from gs_net import add_cat_seg_config, GSNet
import matplotlib.pyplot as plt
from matplotlib.patches import Patch
import glob
import json
import logging


# define constants
CONFIG_PATH = "./configs/vitb_384.yaml"
# CLASS_NAMES_PATH = './datasets/nwpu_discover.json'
# CLASS_NAMES_PATH = './datasets/nwpu_discover_syn.json'
CLASS_NAMES_PATH = './datasets/landdiscover.json'
with open(CLASS_NAMES_PATH, 'r') as f:
    CLASS_NAMES = json.load(f)

# IMG_DIR = "../../datasets/nwpu/pos_img/"
IMG_DIR = "../../datasets/dior/JPEGImages-trainval"
OUTPUT_DIR = "./vis/dior_discover"


MODEL_WEIGHTS_PATH = "../../checkpoints/gsnet/GSNet_base.pth"
DEVICE = "cuda:0"

COLORS = np.random.randint(0, 255, size=(len(CLASS_NAMES), 3), dtype=np.uint8)

def setup_cfg(config_path):
    """Load GSNet config and merge with custom config"""
    logging.getLogger("detectron2").setLevel(logging.WARNING)
    logging.getLogger("fvcore").setLevel(logging.WARNING)
    cfg = get_cfg()
    add_cat_seg_config(cfg)
    cfg.merge_from_file(config_path)
    cfg.MODEL.WEIGHTS = MODEL_WEIGHTS_PATH
    cfg.MODEL.DEVICE = DEVICE
    cfg.freeze()
    default_setup(cfg, None)
    return cfg

def load_model(cfg):
    """加载 GSNet 模型并设置权重"""
    model = GSNet(cfg)
    model.eval()
    model.to(DEVICE)

    # load weights
    checkpointer = DetectionCheckpointer(model)
    checkpointer.load(cfg.MODEL.WEIGHTS)

    return model


def preprocess_image(image_path, cfg):
    """预处理输入图像"""
    # 读取图像
    image = cv2.imread(image_path)
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)  # 转换为 RGB

    # 根据配置文件调整大小
    min_size = cfg.INPUT.MIN_SIZE_TEST  # 测试时的最小尺寸
    height, width = image.shape[:2]
    scale = min_size / min(height, width)
    new_height, new_width = int(height * scale), int(width * scale)
    image = cv2.resize(image, (new_width, new_height), interpolation=cv2.INTER_LINEAR)
    print(f'Resized image from ({height}, {width}) to ({new_height}, {new_width})')
    # 转换为张量
    image_tensor = torch.from_numpy(image.transpose(2, 0, 1)).float()  # HWC -> CHW
    image_tensor = image_tensor.to(DEVICE)

    # 添加 batch 维度
    inputs = [{"image": image_tensor, "height": height, "width": width}]
    return inputs, image



def inference(model, inputs):
    with torch.no_grad():
        predictions = model(inputs)[0]
        print("sem_seg shape:", predictions["sem_seg"].shape)
        print("sem_seg min/max:", predictions["sem_seg"].min(), predictions["sem_seg"].max())
    return predictions

def visualize_results(image, predictions, class_names, image_name, output_dir):
    sem_seg = predictions["sem_seg"].argmax(dim=0).cpu().numpy()
    height, width = image.shape[:2]
    sem_seg_resized = cv2.resize(sem_seg.astype(np.uint8), (width, height), interpolation=cv2.INTER_NEAREST)

    # 创建彩色掩码
    mask = np.zeros_like(image, dtype=np.uint8)
    for idx, class_name in enumerate(class_names):
        mask[sem_seg_resized == idx] = COLORS[idx]

    # 将掩码与原始图像叠加
    blended = cv2.addWeighted(image, 0.5, mask, 0.5, 0.0)

    # 创建图例
    legend_elements = [Patch(facecolor=COLORS[i] / 255.0, edgecolor='k', label=class_names[i]) 
                       for i in range(len(class_names))]

    # 可视化
    plt.figure(figsize=(15, 6))
    plt.subplot(1, 2, 1)
    plt.title("Original Image")
    plt.imshow(image)
    plt.axis("off")
    plt.subplot(1, 2, 2)
    plt.title("Segmentation Result")
    plt.imshow(blended)
    plt.legend(handles=legend_elements, bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.)
    plt.axis("off")
    plt.tight_layout()

    # 保存结果
    os.makedirs(output_dir, exist_ok=True)
    output_path = os.path.join(output_dir, f"{image_name}_seg_result.jpg")
    plt.savefig(output_path, bbox_inches='tight', dpi=300)
    plt.close()

    print(f"Saved segmentation result for {image_name} to {output_path}")


def main():
    cfg = setup_cfg(CONFIG_PATH)
    model = load_model(cfg)

    # 获取图像目录下所有图片，并随机采样 10 张
    image_paths = glob.glob(os.path.join(IMG_DIR, "*.jpg"))
    if not image_paths:
        raise FileNotFoundError(f"No images found in {IMG_DIR}")
    import random
    random.seed(42)
    image_paths = random.sample(image_paths, min(20, len(image_paths)))  # 随机采样，最多 20 张

    for image_path in image_paths:
        # image_path = '/home/zkcs/datasets/dior/JPEGImages-trainval/05660.jpg'
        image_name = os.path.splitext(os.path.basename(image_path))[0]  # 提取文件名，不含扩展名
        print(f"Processing {image_name}...")
        inputs, original_image = preprocess_image(image_path, cfg)
        predictions = inference(model, inputs)
        visualize_results(original_image, predictions, CLASS_NAMES, image_name, OUTPUT_DIR)
    print("Inference completed for all images.")


if __name__ == "__main__":
    main()
```
To perform inference:
```bash
RSIB_CKPT="../../checkpoints/gsnet/RSIB.pth" python inference_gsnet.py
```

**Comment by ChaojunZ** [comment]
> 你好
> 
> 我使用预训练模型在 DIOR 和 NWPU-VHR-10 数据集上测试了一些图像，但我发现该模型无法区分我测试的任何图像中的前景和背景。
> 
> 尽管训练数据集 LandDiscover50k 包括 DIOR 数据集，但性能仍然不尽如人意。
> 
> 您能否对此问题提供一些解释？此外，我还包含了用于推理和可视化的代码。
> 
> 下面提供了示例。
> 
> 感谢您的帮助。
> 
> #### DIOR 数据集中的示例（使用与 LandDiscover 相同的类名）
> ![Image](https://github.com/user-attachments/assets/34356920-efae-4b1c-8e30-1bbe3374f1a8) **00252.jpg**
> 
> ![Image](https://github.com/user-attachments/assets/2dab322f-a664-4268-a572-367386972eb6) **05300.jpg**
> 
> #### examples in NWPU-VHR-10 dataset (use the class names in nwpu dataset with a class)`background`
> ![Image](https://github.com/user-attachments/assets/8dddd239-5b5f-48fe-a7fd-8426641a1aff) **019.jpg**
> 
> ![Image](https://github.com/user-attachments/assets/bfa6bf2c-718b-49da-acd6-33ff60c7a12a) **132.jpg**
> 
> #### Codes inference_gsnet.py
> import os
> import cv2
> import numpy as np
> import torch
> from detectron2.config import get_cfg
> from detectron2.checkpoint import DetectionCheckpointer
> from detectron2.engine import default_setup
> from gs_net import add_cat_seg_config, GSNet
> import matplotlib.pyplot as plt
> from matplotlib.patches import Patch
> import glob
> import json
> import logging
> 
> 
> # define constants
> CONFIG_PATH = "./configs/vitb_384.yaml"
> # CLASS_NAMES_PATH = './datasets/nwpu_discover.json'
> # CLASS_NAMES_PATH = './datasets/nwpu_discover_syn.json'
> CLASS_NAMES_PATH = './datasets/landdiscover.json'
> with open(CLASS_NAMES_PATH, 'r') as f:
>     CLASS_NAMES = json.load(f)
> 
> # IMG_DIR = "../../datasets/nwpu/pos_img/"
> IMG_DIR = "../../datasets/dior/JPEGImages-trainval"
> OUTPUT_DIR = "./vis/dior_discover"
> 
> 
> MODEL_WEIGHTS_PATH = "../../checkpoints/gsnet/GSNet_base.pth"
> DEVICE = "cuda:0"
> 
> COLORS = np.random.randint(0, 255, size=(len(CLASS_NAMES), 3), dtype=np.uint8)
> 
> def setup_cfg(config_path):
>     """Load GSNet config and merge with custom config"""
>     logging.getLogger("detectron2").setLevel(logging.WARNING)
>     logging.getLogger("fvcore").setLevel(logging.WARNING)
>     cfg = get_cfg()
>     add_cat_seg_config(cfg)
>     cfg.merge_from_file(config_path)
>     cfg.MODEL.WEIGHTS = MODEL_WEIGHTS_PATH
>     cfg.MODEL.DEVICE = DEVICE
>     cfg.freeze()
>     default_setup(cfg, None)
>     return cfg
> 
> def load_model(cfg):
>     """加载 GSNet 模型并设置权重"""
>     model = GSNet(cfg)
>     model.eval()
>     model.to(DEVICE)
> 
>     # load weights
>     checkpointer = DetectionCheckpointer(model)
>     checkpointer.load(cfg.MODEL.WEIGHTS)
> 
>     return model
> 
> 
> def preprocess_image(image_path, cfg):
>     """预处理输入图像"""
>     # 读取图像
>     image = cv2.imread(image_path)
>     image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)  # 转换为 RGB
> 
>     # 根据配置文件调整大小
>     min_size = cfg.INPUT.MIN_SIZE_TEST  # 测试时的最小尺寸
>     height, width = image.shape[:2]
>     scale = min_size / min(height, width)
>     new_height, new_width = int(height * scale), int(width * scale)
>     image = cv2.resize(image, (new_width, new_height), interpolation=cv2.INTER_LINEAR)
>     print(f'Resized image from ({height}, {width}) to ({new_height}, {new_width})')
>     # 转换为张量
>     image_tensor = torch.from_numpy(image.transpose(2, 0, 1)).float()  # HWC -> CHW
>     image_tensor = image_tensor.to(DEVICE)
> 
>     # 添加 batch 维度
>     inputs = [{"image": image_tensor, "height": height, "width": width}]
>     return inputs, image
> 
> 
> 
> def inference(model, inputs):
>     with torch.no_grad():
>         predictions = model(inputs)[0]
>         print("sem_seg shape:", predictions["sem_seg"].shape)
>         print("sem_seg min/max:", predictions["sem_seg"].min(), predictions["sem_seg"].max())
>     return predictions
> 
> def visualize_results(image, predictions, class_names, image_name, output_dir):
>     sem_seg = predictions["sem_seg"].argmax(dim=0).cpu().numpy()
>     height, width = image.shape[:2]
>     sem_seg_resized = cv2.resize(sem_seg.astype(np.uint8), (width, height), interpolation=cv2.INTER_NEAREST)
> 
>     # 创建彩色掩码
>     mask = np.zeros_like(image, dtype=np.uint8)
>     for idx, class_name in enumerate(class_names):
>         mask[sem_seg_resized == idx] = COLORS[idx]
> 
>     # 将掩码与原始图像叠加
>     blended = cv2.addWeighted(image, 0.5, mask, 0.5, 0.0)
> 
>     # 创建图例
>     legend_elements = [Patch(facecolor=COLORS[i] / 255.0, edgecolor='k', label=class_names[i]) 
>                        for i in range(len(class_names))]
> 
>     # 可视化
>     plt.figure(figsize=(15, 6))
>     plt.subplot(1, 2, 1)
>     plt.title("Original Image")
>     plt.imshow(image)
>     plt.axis("off")
>     plt.subplot(1, 2, 2)
>     plt.title("Segmentation Result")
>     plt.imshow(blended)
>     plt.legend(handles=legend_elements, bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.)
>     plt.axis("off")
>     plt.tight_layout()
> 
>     # 保存结果
>     os.makedirs(output_dir, exist_ok=True)
>     output_path = os.path.join(output_dir, f"{image_name}_seg_result.jpg")
>     plt.savefig(output_path, bbox_inches='tight', dpi=300)
>     plt.close()
> 
>     print(f"Saved segmentation result for {image_name} to {output_path}")
> 
> 
> def main():
>     cfg = setup_cfg(CONFIG_PATH)
>     model = load_model(cfg)
> 
>     # 获取图像目录下所有图片，并随机采样 10 张
>     image_paths = glob.glob(os.path.join(IMG_DIR, "*.jpg"))
>     if not image_paths:
>         raise FileNotFoundError(f"No images found in {IMG_DIR}")
>     import random
>     random.seed(42)
>     image_paths = random.sample(image_paths, min(20, len(image_paths)))  # 随机采样，最多 20 张
> 
>     for image_path in image_paths:
>         # image_path = '/home/zkcs/datasets/dior/JPEGImages-trainval/05660.jpg'
>         image_name = os.path.splitext(os.path.basename(image_path))[0]  # 提取文件名，不含扩展名
>         print(f"Processing {image_name}...")
>         inputs, original_image = preprocess_image(image_path, cfg)
>         predictions = inference(model, inputs)
>         visualize_results(original_image, predictions, CLASS_NAMES, image_name, OUTPUT_DIR)
>     print("Inference completed for all images.")
> 
> 
> if __name__ == "__main__":
>     main()
> To perform inference:
> 
> RSIB_CKPT="../../checkpoints/gsnet/RSIB.pth" python inference_gsnet.py

你好，请问，你复现这个论文，他的数据集下载，LD50KSplit.zip需要拼接在一起吗。我进行数据的拼接出现了错误。

**Comment by VoyagerXvoyagerx** [comment]
我没有复现论文，只是用了作者公开的权重进行测试

**Comment by ChaojunZ** [comment]
感谢你的回复 使用公开权重 不需要数据集吗

---- Replied Message ----
| From | Yijie ***@***.***> |
| Date | 04/15/2025 21:23 |
| To | yecy749/GSNet ***@***.***> |
| Cc | ChaojunZ ***@***.***>,
Comment ***@***.***> |
| Subject | Re: [yecy749/GSNet] Could not distinguish foreground and background (Issue #4) |

我没有复现论文，只是用了作者公开的权重进行测试

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you commented.Message ID: ***@***.***>

VoyagerXvoyagerx left a comment (yecy749/GSNet#4)

我没有复现论文，只是用了作者公开的权重进行测试

**Comment by chl916185** [comment]
@VoyagerXvoyagerx 
我也是，在测试一些数据后，效果不咋样，您找到合理的解释了吗

**Comment by VoyagerXvoyagerx** [comment]
开放词汇语义分割的方法用在实例分割数据集，似乎都难以区分前景和背景。SegEarth-OV也有这个问题，但比GSNet好一些

**Comment by litingyidalaoban** [comment]
> 开放词汇语义分割的方法用在实例分割数据集，似乎都难以区分前景和背景。SegEarth-OV也有这个问题，但比GSNet好一些

老哥，你指的前景就是需要分割出来的目标词汇对应的目标吗，背景指的是background这个词对应的排除目标之外的区域嘛。

**Comment by VoyagerXvoyagerx** [comment]
> > 开放词汇语义分割的方法用在实例分割数据集，似乎都难以区分前景和背景。SegEarth-OV也有这个问题，但比GSNet好一些
> 
> 老哥，你指的前景就是需要分割出来的目标词汇对应的目标吗，背景指的是background这个词对应的排除目标之外的区域嘛。

yes

**Comment by yecy749** [comment]
Hi, sorry for the late response — I’ve been dealing with some personal matters recently.

Although the model is trained in a supervised manner, it is still based on a CLIP backbone and follows a one-stage open-vocabulary framework. Under this setup, the training process inherently involves a trade-off:

- On one hand, the model needs to retain the open-vocabulary recognition ability of the VLM.
- On the other hand, we try to improve its pixel-level localization ability in the remote sensing domain.

Because the backbone features are not originally optimized for dense prediction, achieving strong performance on the training set is actually very difficult for a CLIP-based OV model. 

This trade-off is also a core challenge in current open-vocabulary segmentation research.
